src/Compilers/CSharp/Portable/Lowering/RuntimeAsyncIteratorRewriter/RuntimeAsyncIteratorRewriter.cs

```diff
@@ -0,0 +1,428 @@
+                EnsureWellKnownMember(WellKnownMember.System_Threading_CancellationTokenSource__Token, bag);
+                EnsureWellKnownMember(WellKnownMember.System_Threading_CancellationTokenSource__Dispose, bag);
+            }
+
+            EnsureWellKnownMember(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__MoveNextAsync, bag);
+            EnsureWellKnownMember(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__get_Current, bag);
+
+            EnsureWellKnownMember(WellKnownMember.System_IAsyncDisposable__DisposeAsync, bag);
+            EnsureWellKnownMember(WellKnownMember.System_Threading_Tasks_ValueTask_T__ctorValue, bag);
+            EnsureWellKnownMember(WellKnownMember.System_Threading_Tasks_ValueTask__ctor, bag);
+
+            ensureSpecialMember(SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__Await_T_FromValueTaskT, bag);
+
+            Symbol ensureSpecialMember(SpecialMember member, BindingDiagnosticBag bag)
+            {
+                return Binder.GetSpecialTypeMember(F.Compilation, member, bag, body.Syntax);
+            }
+        }
+    }
+
+    protected override void GenerateMethodImplementations()
+    {
+        GenerateConstructor();
+
+        if (_isEnumerable)
+        {
+            // IAsyncEnumerable
+            GenerateIAsyncEnumerableImplementation_GetAsyncEnumerator();
+        }
+
+        // IAsyncEnumerator
+        GenerateIAsyncEnumeratorImplementation_MoveNextAsync();
+        GenerateIAsyncEnumeratorImplementation_Current();
+
+        // IAsyncDisposable
+        GenerateIAsyncDisposable_DisposeAsync();
+    }
+
+    private void GenerateConstructor()
+    {
+        Debug.Assert(stateMachineType.Constructor is IteratorConstructor);
+        Debug.Assert(stateField is not null);
+
+        // Produces:
+        // .ctor(int state)
+        // {
+        //     this.state = state;
+        //     this.initialThreadId = {managedThreadId};
+        //     this.instanceId = LocalStoreTracker.GetNewStateMachineInstanceId();
+        // }
+        Debug.Assert(stateMachineType.Constructor is IteratorConstructor);
+
+        F.CurrentFunction = stateMachineType.Constructor;
+        var blockBuilder = ArrayBuilder<BoundStatement>.GetInstance();
+        blockBuilder.Add(F.BaseInitialization());
+
+        blockBuilder.Add(F.Assignment(F.InstanceField(stateField), F.Parameter(F.CurrentFunction.Parameters[0]))); // this.state = state;
+
+        BoundExpression managedThreadId = MakeCurrentThreadId();
+        if (initialThreadIdField is not null)
+        {
+            // this.initialThreadId = {managedThreadId};
+            blockBuilder.Add(F.Assignment(F.InstanceField(initialThreadIdField), managedThreadId));
+        }
+
+        if (instanceIdField is not null &&
+            F.WellKnownMethod(WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__GetNewStateMachineInstanceId) is { } getId)
```

Comment:
Can this ever return false when `instanceIdField` is `not null`? If so, what happens?